File: c-004/main.go
package main

import (
	"math"
	"strconv"
)

func IsPalindrome(num int) bool {
	str := strconv.FormatInt(int64(num), 10)
	for i := 0; i < len(str)/2; i++ {
		if str[i] != str[len(str)-i-1] {
			return false
		}
	}
	return true
}

func LargestPalindrome(num int) int {
	if num == 0 {
		return 0
	}
	result := 0
	for i := num; i > 0; i-- {
		if IsPalindrome(num * i) {
			result = num * i
			break
		}
	}
	return int(math.Max(float64(result), float64(LargestPalindrome(num-1))))
}

func main() {
	println(LargestPalindrome(999))
}

File: c-004/main.rb
def isPalindrome num
    str = num.to_s
    str == str.reverse
end

def largestPalindrome num
    result = 0
    for i in (0..num).reverse_each do
        if isPalindrome(num * i) then
            result = num * i
            break
        end
    end
    num > 0 ? [result, largestPalindrome(num-1)].max : 0
end

puts largestPalindrome(999)
